### ISSUE TITLE
build error: unknown type name 'locale_t' on Solaris 10

### ISSUE THREAD

**Issue by mojca** [fontforge/fontforge#3027]
When I build fontforge as part of another software (@mgieseki's dvisvgm) on Solaris 10 (i386) with gcc 5.2 (in c++11 mode) I get the following error:

```
fontforge/splinefont.h:2569:39: error: unknown type name 'locale_t'
 static inline void switch_to_c_locale(locale_t * tmplocale_p, locale_t * oldlocale_p) {
                                       ^
```

This seems related to #1717 and #1983. I'm not sure if the issue with cygwin has been solved so far, but the modification suggested in that bug report has been applied in the meantime.

I managed to work around the issue by applying the following patch:

```patch
--- fontforge/splinefont.h.orig
+++ fontforge/splinefont.h
@@ -2554,7 +2554,7 @@
 
 #include "ustring.h"
 
-#ifdef _WIN32
+#if 1
 #define BAD_LOCALE_HACK
 typedef char* locale_t;
 #define LC_GLOBAL_LOCALE ((locale_t)-1)
```

The fun part is that the same sources apparently work just fine on Solaris 11, so doing just an "if this is windows or solaris" doesn't sound like a good idea.

It later failed at linking stage (but that might be a problem elsewhere):
```
Undefined                       first referenced
 symbol                             in file
libintl_gettext                 ../libs/ff-woff/libfontforge.a(nowakowskittfinstr.o)
ld: fatal: symbol referencing errors. No output written to dvisvgm
```

I'm not sure which version of fontforge is being used, I wasn't able to find the version information, but you can check it [here](https://github.com/mgieseki/dvisvgm/tree/master/libs/ff-woff) and I will try to build fontforge from your upstream sources next.

### When reporting a bug/issue:
- [ ] Screenshot
- [ ] The FontForge version and the operating system you're using
- [ ] The behavior you expect to see, and the actual behavior
- [ ] Steps to reproduce the behavior 
- [ ] \(optional) Possible solution/fix/workaround

**Comment by mojca** [comment]
I tried to compile the original sources, but it's not yet clear to me how.

I tried fetching the git sources first, but I was not able to bootstrap it on Solaris (maybe I could bootstrap it elsewhere and then move the whole folder there). It failed if there was no internet access in the build environment and apparently had some strict idea about what tools had to be present for bootstrapping.

So I downloaded [fontforge-20161004.tar.gz](https://github.com/fontforge/fontforge/releases/download/20161005/fontforge-20161004.tar.gz), but running `./configure` doesn't seem to work:
```
$ ./configure 
configure: error: cannot find install-sh, install.sh, or shtool in config "."/config
```
(and the tarball contains lots of git repositories as well as actual `Makefile`s).

**Comment by jtanx** [comment]
It looks like Solaris doesn't have [uselocale](https://www.gnu.org/software/gnulib/manual/html_node/uselocale.html), so the patch you have above is probably correct for Solaris. I'm not too sure about the second error, but perhaps there's something up with gettext/libintl

Try using [fontforge-dist-20161004.tar.gz](https://github.com/fontforge/fontforge/releases/download/20161005/fontforge-dist-20161004.tar.gz)

**Comment by mojca** [comment]
Thank you, it builds now. Just curious, should I worry about the following warning?
```
In file included from ../../fontforge-2.0.20161004/inc/basics.h:30:0,
                 from ../../fontforge-2.0.20161004/inc/utype.h:10,
                 from ../../fontforge-2.0.20161004/Unicode/ArabicForms.c:7:
../inc/fontforge-config.h:1507:0: warning: "_FILE_OFFSET_BITS" redefined
 #define _FILE_OFFSET_BITS 64
 ^
In file included from /usr/include/iso/ctype_iso.h:30:0,
                 from /usr/include/ctype.h:18,
                 from ../../fontforge-2.0.20161004/inc/utype.h:9,
                 from ../../fontforge-2.0.20161004/Unicode/ArabicForms.c:7:
/opt/csw/lib/gcc/i386-pc-solaris2.10/5.2.0/include-fixed/sys/feature_tests.h:196:0: note: this is the location of the previous definition
 #define _FILE_OFFSET_BITS 32
 ^
```

**Comment by mojca** [comment]
After the above mentioned patch for "BAD_LOCALE_HACK" the build fails at
```
In file included from ../../fontforge-2.0.20161004/fontforge/fontforge.h:30:0,
                 from ../../fontforge-2.0.20161004/fontforge/ttfspecial.c:27:
../../fontforge-2.0.20161004/fontforge/ttfspecial.c: In function 'ttf_fftm_dump':
../inc/fontforge-config.h:85:31: error: expected expression before '%' token
 #define FONTFORGE_MODTIME_RAW %sL
                               ^
../../fontforge-2.0.20161004/fontforge/libffstamp.h:4:24: note: in expansion of macro 'FONTFORGE_MODTIME_RAW'
 #define LibFF_ModTime  FONTFORGE_MODTIME_RAW /* Seconds since 1970 (standard unix time) */
                        ^
../../fontforge-2.0.20161004/fontforge/ttfspecial.c:2229:22: note: in expansion of macro 'LibFF_ModTime'
     cvt_unix_to_1904(LibFF_ModTime,results);
                      ^
../../fontforge-2.0.20161004/fontforge/ttfspecial.c:2229:5: error: too few arguments to function 'cvt_unix_to_1904'
     cvt_unix_to_1904(LibFF_ModTime,results);
     ^
In file included from ../../fontforge-2.0.20161004/fontforge/ttfspecial.c:34:0:
../../fontforge-2.0.20161004/fontforge/ttf.h:855:13: note: declared here
 extern void cvt_unix_to_1904( long long time, int32 result[2]);
             ^
```
so I'm unable to proceed with testing anything else.

**Comment by jtanx** [comment]
It looks like your `date` executable doesn't accept the `%s` modifier. The output of:

```bash
date +'%s'
```

Should be something like 1486764782 which I think is the unix epoch. If you run `date --version`,  `date --help` and `man date` and post its output, we could probably figure out how to set it correctly.

You can use this script to test the output:

```bash
#!/bin/bash

FONTFORGE_DATE_NOW=${SOURCE_DATE_EPOCH:-$(date +'%s')}
FONTFORGE_DATE_OPTS="--date=@${FONTFORGE_DATE_NOW}"
if ! date --version 2>/dev/null ; then
  # Use POSIX date instead of GNU date on BSD systems
  FONTFORGE_DATE_OPTS="-j -f %s ${FONTFORGE_DATE_NOW}"
fi
FONTFORGE_MODTIME="$(date -u $FONTFORGE_DATE_OPTS +'%s')L"
FONTFORGE_MODTIME_STR="$(date -u $FONTFORGE_DATE_OPTS +'%H:%M %Z %e-%b-%Y')"
FONTFORGE_VERSIONDATE="$(date -u $FONTFORGE_DATE_OPTS +'%Y%m%d')"

echo FONTFORGE_DATE_NOW: $FONTFORGE_DATE_NOW
echo FONTFORGE_DATE_OPTS: $FONTFORGE_DATE_OPTS
echo FONTFORGE_MODTIME: $FONTFORGE_MODTIME
echo FONTFORGE_VERSIONDATE: $FONTFORGE_VERSIONDATE
```

Copy and paste that into a file say, `testdate.sh`, then run `chmod +x testdate.sh` to make it executable. Then run `./testdate.sh` - the output should be something like:

```
$./testdate.sh
date (GNU coreutils) 8.25
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David MacKenzie.
FONTFORGE_DATE_NOW: 1486765139
FONTFORGE_DATE_OPTS: --date=@1486765139
FONTFORGE_MODTIME: 1486765139L
FONTFORGE_VERSIONDATE: 20170210
```

----

Once that's figured out, you can either modify `configure.ac` (which is the proper way to do it, but requires you to rerun autoconf, which might not work), or just directly modify `configure`. I'd do the latter for now since you don't have to rerun autoconf. You want to search for and replace:

```bash
FONTFORGE_DATE_NOW=${SOURCE_DATE_EPOCH:-$(date +'%s')}
FONTFORGE_DATE_OPTS="--date=@${FONTFORGE_DATE_NOW}"
if ! date --version 2>/dev/null ; then
  # Use POSIX date instead of GNU date on BSD systems
  FONTFORGE_DATE_OPTS="-j -f %s ${FONTFORGE_DATE_NOW}"
fi
FONTFORGE_MODTIME="$(date -u $FONTFORGE_DATE_OPTS +'%s')L"
FONTFORGE_MODTIME_STR="$(date -u $FONTFORGE_DATE_OPTS +'%H:%M %Z %e-%b-%Y')"
FONTFORGE_VERSIONDATE="$(date -u $FONTFORGE_DATE_OPTS +'%Y%m%d')"
```

with the version that you modified that works

**Comment by mojca** [comment]
```
date: illegal option -- j
date: illegal option -- f
usage:  date [-u] mmddHHMM[[cc]yy][.SS]
        date [-u] [+format]
        date -a [-]sss[.fff]
date: illegal option -- j
date: illegal option -- f
usage:  date [-u] mmddHHMM[[cc]yy][.SS]
        date [-u] [+format]
        date -a [-]sss[.fff]
date: illegal option -- j
date: illegal option -- f
usage:  date [-u] mmddHHMM[[cc]yy][.SS]
        date [-u] [+format]
        date -a [-]sss[.fff]
FONTFORGE_DATE_NOW: %s
FONTFORGE_DATE_OPTS: -j -f %s %s
FONTFORGE_MODTIME: L
FONTFORGE_VERSIONDATE:
```

From `man date`:

```
NAME
     date - write the date and time

SYNOPSIS
     /usr/bin/date [-u] [ +format]

     /usr/bin/date [ -a  [-]sss.fff]

     /usr/bin/date [-u] [ [mmdd] HHMM |  mmddHHMM [cc] yy]  [.SS]

     /usr/xpg4/bin/date [-u] [ +format]

     /usr/xpg4/bin/date [ -a  [-]sss.fff]

     /usr/xpg4/bin/date [-u] [ [mmdd] HHMM |  mmddHHMM  [cc]  yy]
     [.SS]

DESCRIPTION
     The date utility writes the date and time to standard output
     or attempts to set the system date and time. By default, the
     current date and time is written.

     Specifications of native language translations of month  and
     weekday  names  are  supported.  The month and weekday names
     used for a language are based on the locale specified by the
     environment variable LC_TIME. See environ(5).

     The following is the default form for the "C" locale:

     %a %b %e %T %Z %Y


     For example,

     Fri Dec 23 10:10:42 EST 1988
```

**Comment by mojca** [comment]
What I can do is run `gdate`, but that's only part of a package manager and might not generally be available. Then again I highly doubt that one can even remotely compile fontforge without having some package manager installed.

```bash
#!/bin/bash

DATE=gdate

FONTFORGE_DATE_NOW=${SOURCE_DATE_EPOCH:-$($DATE +'%s')}
FONTFORGE_DATE_OPTS="--date=@${FONTFORGE_DATE_NOW}"
if ! $DATE --version 2>/dev/null ; then
  # Use POSIX date instead of GNU date on BSD systems
  FONTFORGE_DATE_OPTS="-j -f %s ${FONTFORGE_DATE_NOW}"
fi
FONTFORGE_MODTIME="$($DATE -u $FONTFORGE_DATE_OPTS +'%s')L"
FONTFORGE_MODTIME_STR="$($DATE -u $FONTFORGE_DATE_OPTS +'%H:%M %Z %e-%b-%Y')"
FONTFORGE_VERSIONDATE="$($DATE -u $FONTFORGE_DATE_OPTS +'%Y%m%d')"

echo FONTFORGE_DATE_NOW: $FONTFORGE_DATE_NOW
echo FONTFORGE_DATE_OPTS: $FONTFORGE_DATE_OPTS
echo FONTFORGE_MODTIME: $FONTFORGE_MODTIME
echo FONTFORGE_VERSIONDATE: $FONTFORGE_VERSIONDATE
```

```
> ./testdate.sh 
date (GNU coreutils) 8.22
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David MacKenzie.
FONTFORGE_DATE_NOW: 1486766173
FONTFORGE_DATE_OPTS: --date=@1486766173
FONTFORGE_MODTIME: 1486766173L
FONTFORGE_VERSIONDATE: 20170210
```

**Comment by mojca** [comment]
An additional remark. Another user who tried to compile FF on [Open Indiana](https://en.wikipedia.org/wiki/OpenIndiana) 11 doesn't need the `BAD_LOCALE_HACK`. Citing him:

> Illumos (i.e., the kernel used by OI) implements this symbol.
> https://illumos.org/man/3c/uselocale

It would probably make sense to use some configure test to check whether this is supported or not. I don't see any clean way to do it right with a bunch of `#ifdef`s.

**Comment by mojca** [comment]
Replacing `date` by `gdate` in `configure` got me much further. Now the next problem seems to be missing support for the `--check` flag in `msgfmt`:

```
/usr/bin/msgfmt --check --output-file=ca.mo ../../fontforge-2.0.20161004/po/ca.po
Usage: msgfmt [-D dir | --directory=dir] [-f | --use-fuzzy]
               [-g] [-o outfile | --output-file=outfile]
               [-s] [--strict] [-v] [--verbose] files ...
```

**Comment by jtanx** [comment]
I'm guessing similar to `gdate`, you need to find the GNU equivalent for msgfmt

**Comment by mojca** [comment]
Oh, there's another one called `gmsgfmt` indeed. But this was less of a problem as removing a single flag was not so problematic anyway.

So basically the list of issues seems to be:
* Properly identify when `BAD_LOCALE_HACK` is needed
* Not sure if that `_FILE_OFFSET_BITS` needs any special treatment?
* Ideally try to run `date` and check if it's compliant; if it's not, try `gdate` and/or allow the user to define the command; I would say that `configure` should already fail if `date` doesn't work as expected (or switch to `gdate`)
* Whatever is needed to get `msgfmt` working out-of-the-box (I see that that one is probably already configurable as `$(MSGFMT)`)

I have a few more issues with fontforge in dvisvgm, but I need to figure out how to reproduce them "in isolation" (maybe I just need to turn the C++11 flags on).

**Comment by JoesCat** [comment]
In terms of _FILE_OFFSET_BITS, I was reading up on this a short while ago, and it has to do with 32bit and 64bit handling (from what I understand). Here's a short bit of info about it (you will see possible errors in section LFS) between 32/64bit results: https://ohse.de/uwe/articles/lfs.html
Font files are fairly small in size, but we are coming into times where you see very large files (example video), and large partitions (Terabyte sizes), so it is something that needs attention.
There is an interesting patch posted here: https://lists.fedorahosted.org/pipermail/elfutils-devel/2015-October/005272.html

This is 'off topic', but in terms of 'date', I'm curious how compatible fontforge/libspiro HEAD is with Solaris, there might be some concepts there worth importing into fontforge, such as the last patch for configure.ac, and some timer stuff in tests/call_test.c (older BSDs see time differently than Linux distros). If it compiles out-of-box, that is a good sign. Likewise, if libuninameslist compiles out-of-box, but to a lesser degree since it doesn't contain any tests in it yet.
...just asking, as it isn't often we hear from other operating systems such as Solaris.

**Comment by mojca** [comment]
I'm not sure I understand the second paragraph. Are you asking me to test https://github.com/fontforge/libspiro and https://github.com/fontforge/libuninameslist?

I'm having some troubles with autotools as well. My main machine would complain about
```
sh: ACLOCAL_FLAGS: command not found
/opt/local/bin/glibtoolize: line 3773: ACLOCAL_FLAGS: command not found
sh: ACLOCAL_FLAGS: command not found
```
while Solaris would not be able to generate the files at all.
```
$ autoreconf -i
sh: syntax error at line 1: `(' unexpected
autoreconf: /opt/csw/bin/aclocal failed with exit status: 2

$ autoconf -i
configure.ac:39: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:52: error: possibly undefined macro: AC_ENABLE_SHARED
configure.ac:146: error: possibly undefined macro: AM_CONDITIONAL
configure.ac:238: error: possibly undefined macro: AC_PROG_LIBTOOL
```

**Comment by JoesCat** [comment]
Thanks - that was informative.

autoreconf and automake are not normally out-of-box tools, they are usually in the hands of developers.

I've updated the release tags and added [libuninameslist-dist-20160701.tar.gz](https://github.com/fontforge/libuninameslist/releases/download/20160701/libuninameslist-dist-20160701.tar.gz) and [libspiro-dist-0.5.20150702.tar.gz](https://github.com/fontforge/libspiro/releases/download/0.5.20150702/libspiro-dist-0.5.20150702.tar.gz). Both of these (I hope) should run without needing to do autoreconf -i; automake since they already are prepared for ./configure; make.
...or that's how I hope it goes...

As a user, these normally install into /usr/local by default.
Fontforge "should" be able to locate them, and you should see them listed in the results when you run FontForge ./configure
If not found, then this also indicates that Solaris has trouble searching for user installed libraries. Some distros make it easy to also search /usr/local while other distros make it harder, or are not quite up to date in this area. If Solaris does not find these, it's enough to tell me Solaris has problems in this area too, but if you want to go further, then you would need to re-install them in /usr
./configure --prefix=/usr
make
su
make install
exit
in which case FonForge ./configure should find them (listed in results).
Ideally, you shouldn't need to go this far.
Neither of these two libraries are needed for dvisvgm, but whatever results/problems you see installing them are helpful - Thanks.

**Comment by mojca** [comment]
I cannot afford to install anything to `/usr` or `/usr/local`, the machine is not mine, but I guess you asked me to check whether the software builds at all? Or do you want me to also check if fontforge is happy with those two libraries? (In that case I should be able to point fontforge to a private, arbitrary installation of the libraries, I guess.)

**Comment by mojca** [comment]
`libspiro` builds and passes all the tests, both with gcc 5.2 and the Sun's compiler. With gcc I get:
```
In file included from ../libspiro-0.5.20150702/spiro.c:31:0:
../libspiro-0.5.20150702/spiro.c: In function 'check_finiteness':
./spiro-config.h:93:22: warning: implicit declaration of function 'finite' [-Wimplicit-function-declaration]
 #define IS_FINITE(x) finite(x)
                      ^
../libspiro-0.5.20150702/spiro.c:843:11: note: in expansion of macro 'IS_FINITE'
      if ( IS_FINITE( segs[i].ks[j])==0 ) return -1;
           ^
```
and with `cc` I get a similar warning
```
"../../libspiro-0.5.20150702/spiro.c", line 843: warning: implicit function declaration: finite
```

`libuninameslist` also builds with both compilers, no idea how to test anything.

Sun's compiler complained about
```
cc: Warning: illegal option -fno-common
```

**Comment by JoesCat** [comment]
On February 12, 2017 12:51:51 AM Mojca Miklavec wrote:
> I cannot afford to install anything to `/usr` or `/usr/local`, the
> machine is not mine, but I guess you asked me to check whether the
> software builds at all?


If you cannot install them, no worries, the information you provided now is 
plenty good, and provides confidence that it is about as good as can be 
done. I was a little concerned - thanks for checking these with Solaris.

> Or do you want me to also check if fontforge is
> happy with those two libraries? (In that case I should be able to point
> fontforge to a private, arbitrary installation of the libraries, I
> guess.)


Both libraries are more of interest for font developers. If you aren't 
developing fonts, it is unlikely needed, however, the fact that you were 
able to build them (and bonus run all tests for libspiro) is good info.

If you are interested for your own curiosity on what they do, the libspiro 
library helps with doing spiros:
http://designwithfontforge.com/en-US/Drawing_With_Spiro.html

and libuninameslist provides names and added information when you hover the 
mouse over the characters - mainly of interest to font developers. I did 
not add any tests for this library yet, the original library was a large 
array of data, but now that it has some runnable functions, I should add 
some tests for it too.

I'll need to investigate "finite(x)", I've seen it before and it's a macro 
in the math.h library. configure seems to have problems with macros.

Thanks for this side-trip.

Going back to your original thread, and searching around:
Whatever is needed to get msgfmt working out-of-the-box (I see that that 
one is probably already configurable as $(MSGFMT))

Are you able to install gettext?

**Comment by mgieseki** [comment]
> I'll need to investigate "finite(x)", I've seen it before and it's a macro in the math.h library.

I had a similar issue when building libfontforge for dvisvgm. `finite(x)` is not part of ANSI C, so some compilers don't support it. As of C99 there is `isfinite(x)` in `math.h` which I use in the bundled fontforge library for dvisvgm. However, that's probably not an option for the original fontforge sources.

**Comment by JoesCat** [comment]
On February 12, 2017 03:02:35 AM Martin Gieseking wrote:
> > I'll need to investigate "finite(x)", I've seen it before and it's a
> > macro in the math.h library.
> 
> I had a similar issue when building libfontforge for dvisvgm.
> `finite(x)` is not part of ANSI C, so some compilers don't support it.
> As of C99 there is `isfinite(x)` in `math.h` which I use in the bundled
> fontforge library for dvisvgm. However, that's probably not an option
> for the original fontforge sources.

in libspiro/confonfigure.ac there is an function that tests for finite and 
then is_finite, can you switch the two terms and try that on libspiro?
We will need to implement the same into fontforge.

**Comment by mojca** [comment]
> Are you able to install gettext?

Yes. (OK, what I really want to do is to compile a portable binary for which I don't want to have a dependency on a shared library that won't be present on the target computer, but that's a different issue.)

The failure with linking against gettext is something weird that I still need to properly debug/diagnose. In any case I don't run into that problem when compiling fontforge alone, it's just present when I compile it as a library as part of dvisvgm. I could write more, but that only makes sense after I dive a bit deeper into the issue.

The only thing about getting gettext work out-of-the-box would be some configure-time check which would test for support of the `--check` flag for `msgfmt` upfront (or perhaps even an option that would allow me to disable gettext support altogether). At the moment I can live with setting `$(MSGFMT)` manually.

The higher priority would be to avoid having to manually fiddle with `date` and `BAD_LOCALE_HACK`.

**Comment by JoesCat** [comment]
just adding another time reference: https://sourceforge.net/p/fontforge/mailman/message/36280635/

**Comment by skef** [comment]
Since you're "around" @mojca -- should we leave this issue open?

**Comment by mojca** [comment]
Meanwhile the full suite for which I needed this and on which I run CI on regular basis compiles even on Sparc Solaris 10, but it uses a somewhat stripped-down version of fontforge (and potentially not updated since a very long time), so it's not a super reliable reference.

I checked the source code and the original lines which I had to patch are no longer there either.

The warning about FILE_OFFSET_BITS redefined to a different value (32 vs. 64) is still there.

I'm not sure if I'm super thrilled to try to do the full round with trying to do a standalone build on Solaris, in particular due to a large number of dependencies, bootstrapping etc., but the original problem I had was apparently solved in the meantime (or at least the tiny bit which concerned me).
